src/context/AppProvider/AppContext.tsx

```diff
@@ -192,64 +190,121 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
+
+              const accountNumberHasChainActivity = !isUtxoChainId(chainId)
+                ? hasActivity
+                : // For UTXO AccountIds, we need to check if *any* of the scriptTypes have activity, not only the current one
+                  // else, we might end up with partial account data, with only the first 1 or 2 out of 3 scriptTypes
+                  // being upserted for BTC and LTC
+                  results.some((res, _idx) => {
+                    if (res.status === 'rejected') return false
+                    const { data: account } = res.value
+                    if (!account) return false
+                    const accountId = accountIds[_idx]
+                    const { chainId: _chainId } = fromAccountId(accountId)
+                    if (chainId !== _chainId) return false
+                    return account.accounts.byId[accountId].hasActivity
+                  })
+
+              // don't add accounts with no activity past account 0
+              if (accountNumber > 0 && !accountNumberHasChainActivity) {
+                chainIdsWithActivity = chainIdsWithActivity.filter(_chainId => _chainId !== chainId)
+                delete accountMetadataByAccountId[accountId]
+              } else {
+                // unique set to handle utxo chains with multiple account types per account
+                chainIdsWithActivity = Array.from(new Set([...chainIdsWithActivity, chainId]))
+
+                dispatch(portfolio.actions.upsertPortfolio(account))
+                const chainIdAccountMetadata = Object.entries(accountMetadataByAccountId).reduce(
+                  (acc, [accountId, metadata]) => {
+                    const { chainId: _chainId } = fromAccountId(accountId)
+                    if (chainId === _chainId) {
+                      acc[accountId] = metadata
+                    }
+                    return acc
+                  },
+                  {} as AccountMetadataById,
+                )
+                dispatch(
+                  portfolio.actions.upsertAccountMetadata({
+                    accountMetadataByAccountId: chainIdAccountMetadata,
+                    walletId,
+                  }),
+                )
+                for (const accountId of Object.keys(accountMetadataByAccountId)) {
+                  dispatch(portfolio.actions.enableAccountId(accountId))
+                }
+              }
+            })
+
+            return results
           })
 
-          chainIds = chainIdsWithActivity
-        }
+          await Promise.allSettled(accountNumberAccountIdsPromises)
 
-        dispatch(
-          portfolio.actions.upsertAccountMetadata({
-            accountMetadataByAccountId,
-            walletId: await wallet.getDeviceID(),
-          }),
-        )
-
-        for (const accountId of Object.keys(accountMetadataByAccountId)) {
-          dispatch(portfolio.actions.enableAccountId(accountId))
+          chainIds = chainIdsWithActivity
```

Comment:
[suggestion]: For both performance and readability, `chainIdsWithActivity` would be much better suited as a `const chainIdsWithActivity: Set<ChainId> = new Set()` instead of a `let chainIdsWithActivity: string[] = []`. This way, above you can mutate the set rather than reassigning a new array at every step:

```typescript
chainIdsWithActivity.delete(chainId)
chainIdsWithActivity.add(chainId)
```

instead of 

```typescript
chainIdsWithActivity = chainIdsWithActivity.filter(_chainId => _chainId !== chainId)
chainIdsWithActivity = Array.from(new Set([...chainIdsWithActivity, chainId]))
```

and then at the end you can assign `chainIds = Array.from(chainIdsWithActivity)`

Reply:
Much cleaner indeed! Went with keeping everything as a set and doing in-place `Array.from()` for the one place where `chainIds` is required by another API (`deriveAccountIdsAndMetadata`) to be an array: 

81b6e9b40fc813fcd56b424bd355dfdad66249f1 and confirmed we're still happy! https://jam.dev/c/6625fa90-5f30-4cd1-9acb-b46a9ad82dad